src/file-manager/fm-desktop-wayland-bg-dialog.c

```diff
@@ -0,0 +1,454 @@
+
+        preview_image = gtk_image_new ();
+        provider = gtk_css_provider_new ();
+        gtk_style_context_add_provider (gtk_widget_get_style_context (preview),
+                                        GTK_STYLE_PROVIDER (provider),
+                                        GTK_STYLE_PROVIDER_PRIORITY_APPLICATION);
+        gtk_box_pack_start (GTK_BOX (preview), preview_image, FALSE, FALSE, 0);
+        gtk_box_pack_start (GTK_BOX (box), preview, FALSE, FALSE, 0);
+    }
+
+    /* No image filename means we are previewing a color or gradient background*/
+    if ((!filename) || (strcmp (filename, "") == 0) || (strcmp (filename, " ") == 0))
+
+    {
+        gtk_image_clear (GTK_IMAGE(preview_image));
+
+        /*Build a color preview using a cssprovider due to requirement to handle RBGA values*/
+        string = g_string_new(NULL);
+        g_string_append (string, "#caja-wayland-bg-preview {");
+
+        if (strcmp (shading_type, "vertical-gradient") == 0)
+            g_string_append (string, "background-image: linear-gradient(to bottom,");
+
+        else if  (strcmp (shading_type, "horizontal-gradient") == 0)
+            g_string_append (string, "background-image: linear-gradient(to right,");
+
+        else
+        {
+            g_string_append (string, "background-color:");
+            g_string_append (string, primary_color_str);
+        }
+
+        if ((strcmp (shading_type, "vertical-gradient") == 0) ||
+           (strcmp (shading_type, "horizontal-gradient") == 0))
+        {
+            g_string_append (string, primary_color_str);
+            g_string_append (string, ",");
+            g_string_append (string, secondary_color_str);
+            g_string_append (string, ");");
+        }
+            g_string_append (string, "}");
+
+        css = g_string_free (string, FALSE);
+
+        gtk_css_provider_load_from_data (provider, css, -1, NULL);
+
+        g_free (css);
+
+    }
+    else
+    /*Preview a background image*/
+    {
+        GdkPixbuf  *pixbuf;
+
+        pixbuf = gdk_pixbuf_new_from_file_at_scale (filename, geometry.width / 5,
+                                                    geometry.height / 5, TRUE, NULL);
+
+        gtk_image_set_from_pixbuf (GTK_IMAGE (preview_image), pixbuf);
+
+
+        /*Clear the color preview*/
+        string = g_string_new (NULL);
+        g_string_append (string, "#caja-wayland-bg-preview {");
+
+        g_string_append (string, "background-image: none;");
+        g_string_append (string, "background-color: transparent;");
+        g_string_append (string, "}");
+
+        css = g_string_free (string, FALSE);
+        gtk_css_provider_load_from_data (provider, css, -1, NULL);
+
+        g_free (css);
+        g_object_unref (pixbuf);
+    }
+}
+
+static void
+update_primary_color (GtkWidget *colorbutton1)
+{
+    const gchar *shading_type, *primary_color_str, *secondary_color_str;
+    GdkRGBA color1;
+
+    shading_type = g_settings_get_string (background_settings,
```

Comment:
Memory should be released ```g_free (shading_type)```

Reply:
Freed in line 178 of the new PR version of this file